Implement caretPositionFromPoint()

rwlbuis opened the pull request:
#### 129140715fbb75dd8c82cf598f622c98deee56ba
<pre>
Implement caretPositionFromPoint()
<a href="https://bugs.webkit.org/show_bug.cgi?id=172137">https://bugs.webkit.org/show_bug.cgi?id=172137</a>

Reviewed by Darin Adler.

Implement Document.caretPositionFromPoint() based on the latest draft:
<a href="https://drafts.csswg.org/cssom-view-1/#dom-document-caretpositionfrompoint">https://drafts.csswg.org/cssom-view-1/#dom-document-caretpositionfrompoint</a>

The feature depends on the runtime flag CaretPositionFromPointEnabled, disabled by default.

This PR imports shadow-dom/Document-caretPositionFromPoint.tentative.html to test
the shadowRoots argument.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/caretPositionFromPoint-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/caretPositionFromPoint-with-transformation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/Document-caretPositionFromPoint.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/Document-caretPositionFromPoint.tentative.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/CaretPosition.cpp: Copied from Source/WebCore/dom/InternalObserver.cpp.
(WebCore::CaretPosition::CaretPosition):
(WebCore::CaretPosition::getClientRect):
* Source/WebCore/dom/CaretPosition.h: Copied from Source/WebCore/dom/InternalObserver.cpp.
(WebCore::CaretPosition::create):
(WebCore::CaretPosition::offsetNode const):
(WebCore::CaretPosition::offset const):
* Source/WebCore/dom/CaretPosition.idl: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/CaretPositionFromPointOptions.h: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/CaretPositionFromPointOptions.idl: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/DeviceMotionEvent.cpp:
* Source/WebCore/dom/Document+CaretPositionFromPoint.idl: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::caretPositionFromPoint):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentOrShadowRoot.idl:
* Source/WebCore/dom/InternalObserver.cpp:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::getBoundingClientRect const):
(WebCore::Range::getBoundingClientRect):
* Source/WebCore/dom/Range.h:

Canonical link: <a href="https://commits.webkit.org/287047@main">https://commits.webkit.org/287047@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c072d0f1aad6701b4617220569f80fc994563629

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57126 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31460 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82741 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29346 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80214 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66280 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5413 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61154 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19072 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81153 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51136 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/67730 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41467 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24656 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27686 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71229 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69590 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25009 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84102 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77321 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5452 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3682 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69375 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5608 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67023 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68629 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17126 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12623 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10839 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99634 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5400 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/8153 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21770 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5390 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8822 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7178 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1868,6 +1868,20 @@ CaretBrowsingEnabled:
+  type: bool
+  status: testable
+  category: dom
+  humanReadableName: "CaretPositionFromPoint API"
```
Shouldn't this be `document.caretPositionFromPoint()` or some such?

rwlbuis replied:
Yes, that is more specific, fixed.

annevk reviewed (commented):
I'm probably not the best person to review this. I'm not sufficiently familiar with WebKit's hit testing code.

darinadler commented on Source/WebCore/dom/CaretPosition.cpp:
```diff
@@ -0,0 +1,52 @@
+        return nullptr;
+
+    auto range = createLiveRange(SimpleRange { BoundaryPoint { *m_offsetNode, m_offset }, BoundaryPoint { *m_offsetNode, m_offset } });
+    return range->getBoundingClientRect();
```
We could easily refactor `Range::getBoundingClientRect` to call something that takes a `SimpleRange` since that’s what it does. There’s no need to create and destroy a live range here. While it’s not the worst thing, I would like to avoid creating any live ranges unless we absolutely have to.

rwlbuis replied:
I tried to do a refactor.

darinadler commented on Source/WebCore/dom/CaretPosition.h:
```diff
@@ -0,0 +1,56 @@
+#include "ScriptWrappable.h"
+#include <wtf/Ref.h>
+#include <wtf/RefCounted.h>
+#include <wtf/TZoneMalloc.h>
```
Too many includes here. Should be just `ContainerNode.h` and `ScriptWrappable.h`. And could also get rid of the include of `ContainerNode.h` if we move the destructor definition into the .cpp file and add some forward declarations.

rwlbuis replied:
True, not all of them are needed, fixed.

darinadler commented on Source/WebCore/dom/CaretPosition.h:
```diff
@@ -0,0 +1,56 @@
+    RefPtr<DOMRect> getClientRect();
+
+protected:
+    CaretPosition(RefPtr<Node>&& offsetNode, unsigned offset);
```
This should be private.

rwlbuis replied:
Done.

darinadler commented on Source/WebCore/dom/CaretPosition.idl:
```diff
@@ -0,0 +1,34 @@
+] interface CaretPosition {
+  readonly attribute Node offsetNode;
+  readonly attribute unsigned long offset;
+  [NewObject] DOMRect? getClientRect();
```
For WebKit coding style should be four space indented?

rwlbuis replied:
Right, in my defense I am not adding idl files every day :) Fixed.

darinadler commented on Source/WebCore/dom/CaretPositionFromPointOptions.idl:
```diff
@@ -0,0 +1,28 @@
+ */
+
+dictionary CaretPositionFromPointOptions {
+  sequence<ShadowRoot> shadowRoots = [];
```
For WebKit coding style should be four space indented?

rwlbuis replied:
Fixed.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2097,6 +2099,54 @@ std::optional<BoundaryPoint> Document::caretPositionFromPoint(const LayoutPoint&
+    if (!hasLivingRenderTree())
+        return nullptr;
+
+    LayoutPoint clientPoint = { x, y };
```
Not sure we need the `=` here.

rwlbuis replied:
Yes, wasteful, in fact we do not need a named variable at all (originally I was probably thinking it would be used in more than one place).

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2097,6 +2099,54 @@ std::optional<BoundaryPoint> Document::caretPositionFromPoint(const LayoutPoint&
+
+    RefPtr anchorNode = position.anchorNode();
+    if (anchorNode->shadowHost()) {
+        if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(*anchorNode->shadowHost())) {
```
```suggestion
    if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(anchorNode->shadowHost())) {
```
Leaving out the `*` allows `dynamicDowncast` to do the null check for us.

rwlbuis replied:
Clever, fixed.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2097,6 +2099,54 @@ std::optional<BoundaryPoint> Document::caretPositionFromPoint(const LayoutPoint&
+    position = position.parentAnchoredEquivalent();
+
+    anchorNode = position.anchorNode();
+    return CaretPosition::create(WTFMove(anchorNode), position.offsetInContainerNode());
```
```suggestion
    return CaretPosition::create(position.anchorNode(), position.offsetInContainerNode());
```

rwlbuis replied:
Done.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2097,6 +2099,54 @@ std::optional<BoundaryPoint> Document::caretPositionFromPoint(const LayoutPoint&
+    if (anchorNode->shadowHost()) {
+        if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(*anchorNode->shadowHost())) {
+            anchorNode = textFormControl;
+            if ((!anchorNode->isInShadowTree() && !options.shadowRoots.size()) || (anchorNode->isInShadowTree() && options.shadowRoots.contains(Ref(anchorNode->treeScope().rootNode()))))
```
```suggestion
            if ((!anchorNode->isInShadowTree() && !options.shadowRoots.size()) || (anchorNode->isInShadowTree() && options.shadowRoots.contains(anchorNode->treeScope().rootNode())))
```

rwlbuis replied:
See below.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2097,6 +2099,54 @@ std::optional<BoundaryPoint> Document::caretPositionFromPoint(const LayoutPoint&
+    }
+
+    bool adjustPosition = false;
+    while (anchorNode->isInShadowTree() && !(options.shadowRoots.size() && options.shadowRoots.contains(Ref(anchorNode->treeScope().rootNode())))) {
```
```suggestion
    while (anchorNode->isInShadowTree() && !(options.shadowRoots.size() && options.shadowRoots.contains(anchorNode->treeScope().rootNode()))) {
```

rwlbuis replied:
This unfortunately results in a build error I am not sure how to fix:
app/webkit/Source/WebCore/dom/Document.cpp:2129:140:   required from here
/app/webkit/WebKitBuild/GTK/Release/WTF/Headers/wtf/CheckedRef.h:302:46: error: static assertion failed: Derived class should opt-in when defaulting operator==, or invalid/undefined comparison should be reworked/defined
  302 |         static_assert(defaultedOperatorEqual == DefaultedOperatorEqual::Yes, "Derived class should opt-in when defaulting operator==, or invalid/undefined comparison should be reworked/defined");
      |                       ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/app/webkit/WebKitBuild/GTK/Release/WTF/Headers/wtf/CheckedRef.h:302:46: note: ‘(WTF::DefaultedOperatorEqual::No == WTF::DefaultedOperatorEqual::Yes)’ evaluates to false
[1202/2373] Building CXX object Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-be65d27a-17.cpp.o

darinadler replied:
OK. A bit mysterious but not worth major effort I guess.

darinadler commented on Source/WebCore/dom/InternalObserver.cpp:
```diff
@@ -28,8 +28,9 @@
 #include "JSDOMExceptionHandling.h"
+#include <JavaScriptCore/Exception.h>
 #include <JavaScriptCore/JSCJSValue.h>
-#include <JavaScriptCore/JSGlobalObjectInlines.h>
```
Why isn’t this needed any more? Seems really surprising this file doesn’t call any functions that have inline bodies in that header.

rwlbuis replied:
It looks like we can just include JSGlobalObject.h instead of JSGlobalObjectInlines.h.

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -272,6 +273,7 @@ class DOMTimerHoldingTank;
 struct ApplicationManifest;
 struct BoundaryPoint;
+struct CaretPositionFromPointOptions;
 struct CSSParserContext;
```
```suggestion
struct CSSParserContext;
struct CaretPositionFromPointOptions;
```
Alphabetical order is case sensitive so we put the capitals first.

rwlbuis replied:
Yes, I was not sure about this one. Seems not caught by style checker. Fixed.

darinadler commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -10631,6 +10637,8 @@
 		447958021643B47B001E0A7F /* ParsedContentType.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = ParsedContentType.cpp; sourceTree = "<group>"; };
 		447958031643B47B001E0A7F /* ParsedContentType.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = ParsedContentType.h; sourceTree = "<group>"; };
+		447B137C2CB6B80C00F355CB /* JSCaretPositionFromPointOptions.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; name = JSCaretPositionFromPointOptions.h; path = "/Users/rbuis/Library/Developer/Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPositionFromPointOptions.h"; sourceTree = "<absolute>"; };
+		447B137D2CB6B80C00F355CB /* JSCaretPositionFromPointOptions.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; name = JSCaretPositionFromPointOptions.cpp; path = "/Users/rbuis/Library/Developer/Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPositionFromPointOptions.cpp"; sourceTree = "<absolute>"; ...
```
Can’t check this in with paths that are absolute and include your user name! Needs to be changed to be group-relative.

rwlbuis replied:
Yes, that was sloppy, should have checked. Fixed.

darinadler commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -10675,6 +10686,8 @@
 		44CF5097299B90600038D284 /* RequestPriority.idl */ = {isa = PBXFileReference; lastKnownFileType = text; path = RequestPriority.idl; sourceTree = "<group>"; };
 		44CF5098299B906F0038D284 /* RequestPriority.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = RequestPriority.h; sourceTree = "<group>"; };
+		44D286EB2CAEB9A000E02E84 /* JSCaretPosition.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; name = JSCaretPosition.cpp; path = "/Users/rbuis/Library/Developer/Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPosition.cpp"; sourceTree = "<absolute>"; };
+		44D286ED2CAEB9BA00E02E84 /* JSCaretPosition.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; name = JSCaretPosition.h; path = "/Users/rbuis/Library/Developer/Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPosition.h"; sourceTree = "<absolute>"; };
```
Can’t check this in with paths that are absolute and include your user name! Needs to be changed to be group-relative.

rwlbuis replied:
Fixed.

darinadler commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -10630,6 +10636,8 @@
 		447958021643B47B001E0A7F /* ParsedContentType.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = ParsedContentType.cpp; sourceTree = "<group>"; };
 		447958031643B47B001E0A7F /* ParsedContentType.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = ParsedContentType.h; sourceTree = "<group>"; };
+		447B137C2CB6B80C00F355CB /* JSCaretPositionFromPointOptions.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; name = JSCaretPositionFromPointOptions.h; path = "/Users/rbuis/Library/Developer/Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPositionFromPointOptions.h"; sourceTree = "<absolute>"; };
+		447B137D2CB6B80C00F355CB /* JSCaretPositionFromPointOptions.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; name = JSCaretPositionFromPointOptions.cpp; path = "/Users/rbuis/Library/Developer/Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPositionFromPointOptions.cpp"; sourceTree = "<absolute>"; ...
```
Still need to fix this path to be relative to the group, not absolute.

rwlbuis replied:
Done. Xcode was trying to tell me as much, since the files were marked as red in the file list.

darinadler commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -10634,6 +10640,8 @@
 		447958021643B47B001E0A7F /* ParsedContentType.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = ParsedContentType.cpp; sourceTree = "<group>"; };
 		447958031643B47B001E0A7F /* ParsedContentType.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = ParsedContentType.h; sourceTree = "<group>"; };
+		447B137C2CB6B80C00F355CB /* JSCaretPositionFromPointOptions.h */ = {isa = PBXFileReference; explicitFileType = sourcecode.cpp.h; name = JSCaretPositionFromPointOptions.h; path = "../../../../../../../../Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPositionFromPointOptions.h"; sourceTree = "<group>"; };
+		447B137D2CB6B80C00F355CB /* JSCaretPositionFromPointOptions.cpp */ = {isa = PBXFileReference; explicitFileType = sourcecode.cpp.cpp; name = JSCaretPositionFromPointOptions.cpp; path = "../../../../../../../../Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPositionFromPointOptions.cpp"; sourceTree = "<group>"; };
```
These paths are wrong. Need to be relative to group or intermediates or build products or whatever (match other similar generated files).

rwlbuis replied:
Done.

darinadler commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -10678,6 +10689,8 @@
 		44CF5097299B90600038D284 /* RequestPriority.idl */ = {isa = PBXFileReference; lastKnownFileType = text; path = RequestPriority.idl; sourceTree = "<group>"; };
 		44CF5098299B906F0038D284 /* RequestPriority.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = RequestPriority.h; sourceTree = "<group>"; };
+		44D286EB2CAEB9A000E02E84 /* JSCaretPosition.cpp */ = {isa = PBXFileReference; explicitFileType = sourcecode.cpp.cpp; name = JSCaretPosition.cpp; path = "../../../../../../../../Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPosition.cpp"; sourceTree = "<group>"; };
+		44D286ED2CAEB9BA00E02E84 /* JSCaretPosition.h */ = {isa = PBXFileReference; explicitFileType = sourcecode.cpp.h; name = JSCaretPosition.h; path = "../../../../../../../../Library/Developer/Xcode/DerivedData/WebKit-dvwybpzhkzbzyseogofxedcfxxir/Build/Products/Debug/DerivedSources/WebCore/JSCaretPosition.h"; sourceTree = "<group>"; };
```
These paths are wrong. Need to be relative to group or intermediates or build products or whatever (match other similar generated files).

rwlbuis replied:
Fixed.

darinadler commented on Source/WebCore/dom/Range.h:
```diff
@@ -90,6 +90,7 @@ class Range final : public AbstractRange, public CanMakeSingleThreadWeakPtr<Rang
 
     Ref<DOMRectList> getClientRects() const;
     Ref<DOMRect> getBoundingClientRect() const;
+    static Ref<DOMRect> getBoundingClientRect(const SimpleRange&);
```
Could eventually move this somewhere else since it's not part of the `Range` class any more but it's fine for now. Also could name it using WebKit coding style without the "get" since it's not directly web-exposed.

rwlbuis replied:
I dropped the "get".

darinadler reviewed (changes requested):
Change is good except for Xcode project mistakes.

darinadler commented on Source/WebCore/dom/Range.h:
```diff
@@ -90,6 +90,7 @@ class Range final : public AbstractRange, public CanMakeSingleThreadWeakPtr<Rang
 
     Ref<DOMRectList> getClientRects() const;
     Ref<DOMRect> getBoundingClientRect() const;
+    static Ref<DOMRect> boundingClientRect(const SimpleRange&);
```
We should change this to a non-member function at some point but it's fine like this.

rwlbuis committed:
Implement caretPositionFromPoint()
https://bugs.webkit.org/show_bug.cgi?id=172137

Reviewed by Darin Adler.

Implement Document.caretPositionFromPoint() based on the latest draft:
https://drafts.csswg.org/cssom-view-1/#dom-document-caretpositionfrompoint

The feature depends on the runtime flag CaretPositionFromPointEnabled, disabled by default.

This PR imports shadow-dom/Document-caretPositionFromPoint.tentative.html to test
the shadowRoots argument.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/caretPositionFromPoint-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/caretPositionFromPoint-with-transformation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/Document-caretPositionFromPoint.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/Document-caretPositionFromPoint.tentative.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/CaretPosition.cpp: Copied from Source/WebCore/dom/InternalObserver.cpp.
(WebCore::CaretPosition::CaretPosition):
(WebCore::CaretPosition::getClientRect):
* Source/WebCore/dom/CaretPosition.h: Copied from Source/WebCore/dom/InternalObserver.cpp.
(WebCore::CaretPosition::create):
(WebCore::CaretPosition::offsetNode const):
(WebCore::CaretPosition::offset const):
* Source/WebCore/dom/CaretPosition.idl: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/CaretPositionFromPointOptions.h: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/CaretPositionFromPointOptions.idl: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/DeviceMotionEvent.cpp:
* Source/WebCore/dom/Document+CaretPositionFromPoint.idl: Copied from Source/WebCore/dom/InternalObserver.cpp.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::caretPositionFromPoint):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentOrShadowRoot.idl:
* Source/WebCore/dom/InternalObserver.cpp:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::getBoundingClientRect const):
(WebCore::Range::getBoundingClientRect):
* Source/WebCore/dom/Range.h:

Canonical link: https://commits.webkit.org/287047@main
